harvester_e2e_tests/integration/test_vm_snapshot.py

```diff
@@ -0,0 +1,609 @@
+            break
+        sleep(3)
+    else:
+        raise AssertionError(f"timed out waiting to restore into new VM from "
+                             f"{restored_from_backup_name}")
+
+    yield restored_from_backup_name
+
+    delete_vm(restored_from_backup_name, api_client, wait_timeout)
+
+
+def vm_shell_do(name, api_client, host_shell, vm_shell, user, ssh_keypair, action, wait_timeout):
+    _, privatekey = ssh_keypair
+
+    deadline = datetime.now() + timedelta(seconds=wait_timeout)
+    while deadline > datetime.now():
+        code, data = api_client.vms.get_status(name)
+        if 200 == code:
+            phase = data.get("status", {}).get("phase")
+            conds = data.get("status", {}).get("conditions", [{}])
+            if ("Running" == phase
+                    and "AgentConnected" == conds[-1].get("type")
+                    and data["status"].get("interfaces")):
+                break
+            sleep(3)
+
+        vm_ip = next(iface["ipAddress"] for iface in data["status"]["interfaces"]
+                     if iface["name"] == "default")
+
+        code, data = api_client.hosts.get(data["status"]["nodeName"])
+        host_ip = next(addr["address"] for addr in data["status"]["addresses"]
+                       if addr["type"] == "InternalIP")
+
+        with host_shell.login(host_ip, jumphost=True) as h:
+            vm_sh = vm_shell(user, pkey=privatekey)
+
+            deadline = datetime.now() + timedelta(seconds=wait_timeout)
+            while deadline > datetime.now():
+                try:
+                    vm_sh.connect(vm_ip, jumphost=h.client)
+                except ChannelException as e:
+                    print(e)
+                    sleep(3)
+                else:
+                    break
+            else:
+                raise AssertionError(f"Unable to login to {name}")
+
+            with vm_sh as sh:
+                action(sh)
+
+
+@pytest.mark.p1
+@pytest.mark.virtualmachines
+class TestVMSnapshot:
+    @pytest.mark.dependency(name="source_vm_snapshot")
+    def test_vm_snapshot_create(self, api_client, vm_snapshot):
+        """
+        Test that the VM snapshot can be created.
+
+        Prerequisite:
+        A virtual machine has been created, is running, and has had
+        some well-known test data written to a well-known file and
+        a snapshot has been created.
+
+        The `vm_snapshot` provides all of these prerequisites:
+
+        1. A source image to create the VM from.
+        2. A VM created from the source image.
+        3. A file called `test.txt` containing data: `123`.
+        4. A snapshot called `vm-snapshot`.
+
+        Since all of the heavy lifting is done in the fixture,
+        assert the snapshot was created successfully for this test.
+        """
+        code, data = api_client.vm_snapshots.get(vm_snapshot)
+
+        assert 200 == code
+        assert data.get("status", {}).get("readyToUse") is True
```

Comment:
These checks are already there in the fixture. :) 
Instead of having the logic of creating snapshot and checking the status etc, can we move those steps in the test. We can use the fixtures to create VM etc. The reason for the asked changes are as below:
1. By having the steps in the test will make it more readable.
2. We will isolate the test. As fixture can be used at many places and making a change to only this test will become difficult in future.

Reply:
Incorporated this in the latest revision